// devdecks-image/OptionsMenu.tsx
import * as React from 'react';
import { Button, EditableText, Intent } from '@blueprintjs/core';
import './add-image.scss';

// For select image file dialog prompt
import selectImageFile from './AddImageDialog.tsx'

interface OptionsMenuProps {
  moduleName: string;
  pluginNumber: number;
  pluginState: any;
  slideNumber: number;
  updateCurrentPlugin: Function;
}

class OptionsMenu extends React.Component<OptionsMenuProps, {}> {
  render() {
    const {
      moduleName,
      pluginNumber,
      pluginState,
      slideNumber,
      updateCurrentPlugin
    } = this.props;

    return (
      <div id="devdeck-add-image">
        <h4>Image Options</h4>
        <hr />
        <Button
          id="add-image-button"
          onClick={ () => { selectImageFile(updateCurrentPlugin) } }>
          Upload Image
        </Button>
        <hr />
        <EditableText
          className="add-image-url"
          intent={ Intent.PRIMARY }
          placeholder="Enter URL..."
          value={ pluginState.imageUrl }
          onChange={ (value: string) => updateCurrentPlugin({ imageUrl: value })} />
      </div>
    );
  }
}

export default OptionsMenu;
