// devdecks-textbox/OptionsMenu.tsx
import * as React from 'react';

import FontColor from './options/FontColor';
import FontSize from './options/FontSize';
import FontFamily from './options/FontFamily';
import FontStyles from './options/FontStyles';

interface OptionsMenuProps {
  moduleName: string;
  pluginNumber: number;
  pluginState: any;
  slideNumber: number;
  theme: any;
  addThemeColor: Function;
  updateCurrentPlugin: Function;
}

class OptionsMenu extends React.Component<OptionsMenuProps, {}> {
  render() {
    const {
      moduleName,
      pluginNumber,
      pluginState,
      slideNumber,
      theme,
      addThemeColor,
      updateCurrentPlugin
    } = this.props;

    return (
      <ul id="font-size-options-container" style={{ textAlign: 'center'}}>
        <li>
          <FontSize
            pluginState={ pluginState }
            updateCurrentPlugin={ updateCurrentPlugin } />
        </li>
        <li>
          <FontFamily
            pluginState={ pluginState }
            updateCurrentPlugin={ updateCurrentPlugin } />
        </li>
        <li>
          <FontStyles
            pluginState={ pluginState }
            updateCurrentPlugin={ updateCurrentPlugin } />
        </li>
        <li>
          <FontColor
            pluginState={ pluginState }
            theme={ theme }
            addThemeColor={ addThemeColor }
            updateCurrentPlugin={ updateCurrentPlugin } />
        </li>
      </ul>
    );
  };
}

export default OptionsMenu;
