// devdecks-textbox/TextBox.tsx
import * as React from 'react';
import './text-box.scss';

const ContentEditable = require("react-contenteditable");

interface TextBoxProps {
  disabled: boolean;
  pluginState: any;
  updateCurrentPlugin: Function;
}

const TextBox = ({ disabled, pluginState, updateCurrentPlugin }: TextBoxProps) => {
  const {
    fontBackgroundColor,
    fontColor,
    fontSize,
    fontFamily,
    textAlignment,
    value,
    width,
  } = pluginState;

  return (
    <div
      className="devdecks-textbox-container"
      style={{
        backgroundColor: fontBackgroundColor,
        color: fontColor,
        fontSize: `${fontSize/100}em`,
        fontFamily: fontFamily,
        textAlign: textAlignment,
      }}>
      <ContentEditable
        className="editable-content-editable"
        html={value}
        disabled={disabled}
        onChange={(e: any) => {
          if (!value) {
            updateCurrentPlugin({
              isOrderedList: false,
              isUnorderedList: false,
              value: e.target.value,
            });
          } else {
            updateCurrentPlugin({ value: e.target.value });
          }
        }}
      />
    </div>
  );
}

export default TextBox;
