// devdecks-textbox/options/FontBackgroundColor.tsx
import * as React from 'react';
import { SketchPicker } from 'react-color';
import { Button, Popover, PopoverInteractionKind, Position } from '@blueprintjs/core';

interface FontBackgroundColor {
  pluginState: any;
  slide: any;
  theme: any;
  addThemeColor: Function;
  updateCurrentPlugin: Function;
}

const FontBackgroundColor = ({
  pluginState,
  slide,
  theme,
  addThemeColor,
  updateCurrentPlugin
}: FontBackgroundColor) => {
  const { fontBackgroundColor, fontBackgroundColorHex, fontColor } = pluginState;
  const { colors } = theme;

  const content = (
    <div>
      <SketchPicker
        width='220px'
        color={ fontBackgroundColor }
        presetColors={ colors }
        onChange={ (color) => {
          updateCurrentPlugin({
            fontBackgroundColor: `rgba(${color.rgb.r}, ${color.rgb.g}, ${color.rgb.b}, ${color.rgb.a})`,
            fontBackgroundColorHex: color.hex
          });
        }} />
      <Button
        style={{
          width: '100%',
          margin: '10px auto 0'
        }}
        text="Save Color To Palette"
        onClick={() => addThemeColor(fontBackgroundColorHex)} />
    </div>
  );

  return (
    <Popover content={content}
        interactionKind={PopoverInteractionKind.CLICK}
        position={Position.BOTTOM} >
      <Button
        style={{
          marginTop: 10,
        }}
        text="Background Color" />
    </Popover>
  );
};

export default FontBackgroundColor;
